File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Brute force - every node dfs
func dfs(r *TreeNode, m int)int{
    if r == nil{
        return 0
    }
    res := 0
    if r.Val >= m{
        res = 1
    }
    res += dfs(r.Right, max(m, r.Val))
    res += dfs(r.Left, max(m, r.Val))
    return res
}
func goodNodes(root *TreeNode) int {
    if root == nil{
        return 0
    }
    return dfs(root, root.Val)

}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # max heap of size k? - then k.top - would give kth smallest

        maxheap = [] # size = k

        def rec(root):
            if not root:
                return
            
            v = root.val
            if len(maxheap) < k:
                heapq.heappush(maxheap,-v)
            else:
                if v >= -maxheap[0]:
                    pass
                else:
                    heapq.heappop(maxheap)
                    heapq.heappush(maxheap,-v)
            rec(root.left)
            rec(root.right)
        rec(root)
        return -maxheap[0]

            
            
        


        